# Study Case: D-cycloserine (67)

In [1]:
import mysql.connector
import operator

import numpy as np
import pandas as pd

import datetime

import matplotlib as mlp
import matplotlib.mlab as mlab
import matplotlib.pyplot as plt

#import plotly as py
import plotly.offline as po
import plotly.graph_objs as go
import plotly.express as px

import ipywidgets as widgets
# import seaborn as sns


In [2]:
%matplotlib inline 
po.init_notebook_mode(connected=True)

In [3]:
 pd.set_option("display.max_rows", 5000)                      # To display more rows

In [4]:
 pd.reset_option("^display")                                  # Reset the number of rows displayed to default

We'll use a query so that we can retrieve the metabolites that exist on **bioanalysis**. This is important for eliminating the metabolite duplicates on the following dataframe.

In [5]:
# Retrieve all bioanalysis_metabolite table from DB

cnx = mysql.connector.connect(user='root', password='root',
                              host='127.0.0.1',
                              database='bioeco')

sql_query = pd.read_sql_query(''' select * from bioanalysis_metabolite ''', cnx, parse_dates=True)

df = pd.DataFrame(sql_query, columns=['id','common_name', 'id_bioeco','inchi', 'inchi_key', 'smiles'])

df

id            common_name            id_bioeco  \
0        1             tobramycin  9222799514525203483   
1        2          Telithromycin  9222710493671371933   
2        3        UNII-YCP4CS0W03  9222602442150781005   
3        4     3,5-DIMETHYLPHENOL  9222471406831838542   
4        5     5-Oxooctanoic acid  9222286964594510425   
...    ...                    ...                  ...   
4299  4300         SCHEMBL2580764  7710985295602631400   
4300  4301             Prochloraz  7710932831858002274   
4301  4302      1,2,6-HEXANETRIOL  7710136659027825385   
4302  4303  2-Methylbutanoic acid  7709180902199124149   
4303  4304     (S,S,R,R)-Orlistat  7708370838353571962   

                                                  inchi  \
0     InChI=1S/C18H37N5O9/c19-3-9-8(25)2-7(22)17(29-...   
1     InChI=1S/C43H65N5O10/c1-12-33-43(8)37(48(41(53...   
2     InChI=1S/C6H12N4.H2O4S/c1-4(2)10-6(8)5(7)3-9-1...   
3     InChI=1S/C8H10O/c1-6-3-7(2)5-8(9)4-6/h3-5,9H,1...   
4     InChI=1S/C8H14O3/c1-2-4-7(9)5-3-6-8(10)11/h2-6...   
...                                                 ...   
4299  InChI=1S/C13H23NO3/c1-3-5-6-10(4-2)9-17-13(16)...   
4300  InChI=1S/C15H16Cl3N3O2/c1-2-4-20(15(22)21-5-3-...   
4301    InChI=1S/C6H14O3/c7-4-2-1-3-6(9)5-8/h6-9H,1-5H2   
4302  InChI=1S/C5H10O2/c1-3-4(2)5(6)7/h4H,3H2,1-2H3,...   
4303  InChI=1S/C29H53NO5/c1-5-7-9-11-12-13-14-15-16-...   

                        inchi_key  \
0     NLVFBUXFDBBNBW-PBSUHMDJSA-N   
1     LJVAJPDWBABPEJ-UHFFFAOYSA-N   
2     VKGPVOKDWZAKKE-UHFFFAOYSA-N   
3     TUAMRELNJMMDMT-UHFFFAOYSA-N   
4     OHGVWVVPDOAZSH-UHFFFAOYSA-N   
...                           ...   
4299  YEMWTVRIHVDMPF-UHFFFAOYSA-N   
4300  TVLSRXXIMLFWEO-UHFFFAOYSA-N   
4301  ZWVMLYRJXORSEP-UHFFFAOYSA-N   
4302  WLAMNBDJUVNPJU-UHFFFAOYSA-N   
4303  AHLBNYSZXLDEJQ-GWMMUDDPSA-N   

                                                 smiles  
0     C1C(C(C(C(C1N)OC2C(C(C(C(O2)CO)O)N)O)O)OC3C(CC...  
1     CCC1C2(C(C(C(=O)C(CC(C(C(C(=O)C(C(=O)O1)C)C)OC...  
2                      CC(C)N1C(=C(C=N1)N)N.OS(=O)(=O)O  
3                                    CC1=CC(=CC(=C1)O)C  
4                                     CCCC(=O)CCCC(=O)O  
...                                                 ...  
4299                        CCCCC(CC)COC(=O)C1CCC(=O)N1  
4300       CCCN(CCOC1=C(C=C(C=C1Cl)Cl)Cl)C(=O)N2C=CN=C2  
4301                                      C(CCO)CC(CO)O  
4302                                       CCC(C)C(=O)O  
4303  CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)C(CC(C)...  

[4304 rows x 6 columns]

In [6]:
# Which are the metabolites that are duplicated?

df_met_dupl = df[df.duplicated(['common_name','inchi','inchi_key', 'smiles'], keep = False)]

dupl_met = {}                                           # create a dictionary {metab:[id,id,...]}
for k,g in df_met_dupl.groupby(['common_name'])['id']:
    g = list(g)
    if k not in dupl_met:
        dupl_met[k] = g
    elif k in dupl_met:
        dupl_met[k].extend(g)
        
        
# print(df[df.common_name == 'Glycyrrhizic acid'])   
# the metab 'Glycyrrhizic acid' has 3 ids for the same metabolite but just 2 of them are equal, the other as different InChIKey

print(dupl_met)
    
                

{'4-Benzyloxyphenylacetic acid': [1159, 4062], 'AC1L2SOM': [1204, 3135], 'Creatine phosphate disodium salt': [546, 3567], 'Glycyrrhizic acid': [2592, 3708], 'HYDROXOCOBALAMIN': [2745, 3973], 'L-cystine': [679, 3367], 'Pregna-1,4-diene-3,20-dione,16,21-bis(acetyloxy)-9-fluoro-11,17-dihydroxy-, (11b,16a)-': [3004, 3776], 'SCHEMBL952350': [2945, 4008], 'Temsirolimus': [169, 3514], 'aluminium triphosphate': [3059, 3805], 'deltamethrin': [624, 2567], 'pimaricin': [2974, 4196]}


In [7]:
cnx.close()

Now using other query we'll retrieve the prices for a spcefic period of time **(01/18 a 06/19)**. This query will be the representation of the total number of prices that exist on **bioanalysis** and that have only the **mass units** (g and derivatives).

In [8]:
# Dataframe com query a selecionar todos os preços com unidades de g ou derivados, e no período de tempo entre 01/18 e 06/19

cnx = mysql.connector.connect(user='root', password='root',
                              host='127.0.0.1',
                              database='bioeco')

sql_query = pd.read_sql_query(''' select bioanalysis_metabolite_price.metabolite_id       as metab_id,  
                                  bioanalysis_metabolite.common_name                      as metab_name,
                                  bioanalysis_metabolite_price.date, 
                                  bioanalysis_metabolite_price.price                      as price, 
                                  bioanalysis_metabolite_price.amount                     as amount, 
                                  bioanalysis_metabolite_price.unity                      as unit, 
                                  if(provider.name is null, "None", provider.name)        as provider_name, 
                                  bioanalysis_metabolite_price.source                     as source,
                                  price/amount                                            as price_per_unit,
                                  price/amount*convert_to_g(unity)                        as price_per_g,
                                  year(date)                                              as year, 
                                  month(date)                                             as month
                                  from bioanalysis_metabolite_price
                                  JOIN provider 
                                      on provider.id = bioanalysis_metabolite_price.provider_id
                                  JOIN bioanalysis_metabolite
                                      on bioanalysis_metabolite_price.metabolite_id = bioanalysis_metabolite.id
                                  where convert_to_g(unity) is not null and ((year(date)*10000) + (month(date)*100) between 20180100 and 20190600)
                              ''', cnx, index_col = 'date', parse_dates=True)  

df = pd.DataFrame(sql_query, columns=['metab_id','metab_name', 'price','amount', 'unit', 'provider_name', 'source', 'price_per_unit', 'price_per_g', 'year', 'month'])

df

metab_id                    metab_name  price  amount  \
date                                                                         
2018-01-11 21:39:13         1                    tobramycin  158.0     1.0   
2018-02-11 23:46:53         1                    tobramycin  158.0     1.0   
2018-03-11 23:47:07         1                    tobramycin  158.0     1.0   
2018-04-11 22:47:14         1                    tobramycin  158.0     1.0   
2018-05-11 22:47:27         1                    tobramycin  158.0     1.0   
...                       ...                           ...    ...     ...   
2019-05-06 11:26:13      4128  2-Methyl-5-nitro-2H-indazole  760.0    10.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole   30.0   500.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole   90.0     1.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole  380.0     5.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole  760.0    10.0   

                    unit    provider_name     source  price_per_unit  \
date                                                                   
2018-01-11 21:39:13    g  ACC Corporation  CHEMSPACE          158.00   
2018-02-11 23:46:53    g  ACC Corporation  CHEMSPACE          158.00   
2018-03-11 23:47:07    g  ACC Corporation  CHEMSPACE          158.00   
2018-04-11 22:47:14    g  ACC Corporation  CHEMSPACE          158.00   
2018-05-11 22:47:27    g  ACC Corporation  CHEMSPACE          158.00   
...                  ...              ...        ...             ...   
2019-05-06 11:26:13    g     LeadGen Labs    MolPort           76.00   
2019-06-14 07:36:43   mg     LeadGen Labs    MolPort            0.06   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           90.00   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           76.00   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           76.00   

                     price_per_g  year  month  
date                                           
2018-01-11 21:39:13    158.00000  2018      1  
2018-02-11 23:46:53    158.00000  2018      2  
2018-03-11 23:47:07    158.00000  2018      3  
2018-04-11 22:47:14    158.00000  2018      4  
2018-05-11 22:47:27    158.00000  2018      5  
...                          ...   ...    ...  
2019-05-06 11:26:13     76.00000  2019      5  
2019-06-14 07:36:43      0.00006  2019      6  
2019-06-14 07:36:43     90.00000  2019      6  
2019-06-14 07:36:43     76.00000  2019      6  
2019-06-14 07:36:43     76.00000  2019      6  

[836888 rows x 11 columns]

In [9]:
# To see how many null values exist in each column
df.isnull().sum()

metab_id            0
metab_name        364
price               0
amount              0
unit                0
provider_name       0
source              0
price_per_unit    295
price_per_g       295
year                0
month               0
dtype: int64

In [10]:
df[df.amount == 0]  #there are prices that have the amount equal to zero ...

metab_id                     metab_name  price  amount  \
date                                                                          
2018-10-22 15:22:54        67                  D-cycloserine  105.0     0.0   
2018-11-22 16:23:15        67                  D-cycloserine  105.0     0.0   
2018-12-07 14:41:16        67                  D-cycloserine  105.0     0.0   
2019-01-07 14:41:27        67                  D-cycloserine  105.0     0.0   
2019-02-07 14:41:35        67                  D-cycloserine  105.0     0.0   
...                       ...                            ...    ...     ...   
2018-12-05 19:20:56      4221  Cyclobenzaprine hydrochloride   42.0     0.0   
2019-01-05 19:21:32      4221  Cyclobenzaprine hydrochloride   42.0     0.0   
2019-02-05 19:21:55      4221  Cyclobenzaprine hydrochloride   42.0     0.0   
2019-03-05 19:22:14      4221  Cyclobenzaprine hydrochloride   42.0     0.0   
2019-04-05 18:22:23      4221  Cyclobenzaprine hydrochloride   42.0     0.0   

                    unit provider_name     source  price_per_unit  \
date                                                                
2018-10-22 15:22:54    g     TargetMol  CHEMSPACE             NaN   
2018-11-22 16:23:15    g     TargetMol  CHEMSPACE             NaN   
2018-12-07 14:41:16    g     TargetMol  CHEMSPACE             NaN   
2019-01-07 14:41:27    g     TargetMol  CHEMSPACE             NaN   
2019-02-07 14:41:35    g     TargetMol  CHEMSPACE             NaN   
...                  ...           ...        ...             ...   
2018-12-05 19:20:56    g     TargetMol  CHEMSPACE             NaN   
2019-01-05 19:21:32    g     TargetMol  CHEMSPACE             NaN   
2019-02-05 19:21:55    g     TargetMol  CHEMSPACE             NaN   
2019-03-05 19:22:14    g     TargetMol  CHEMSPACE             NaN   
2019-04-05 18:22:23    g     TargetMol  CHEMSPACE             NaN   

                     price_per_g  year  month  
date                                           
2018-10-22 15:22:54          NaN  2018     10  
2018-11-22 16:23:15          NaN  2018     11  
2018-12-07 14:41:16          NaN  2018     12  
2019-01-07 14:41:27          NaN  2019      1  
2019-02-07 14:41:35          NaN  2019      2  
...                          ...   ...    ...  
2018-12-05 19:20:56          NaN  2018     12  
2019-01-05 19:21:32          NaN  2019      1  
2019-02-05 19:21:55          NaN  2019      2  
2019-03-05 19:22:14          NaN  2019      3  
2019-04-05 18:22:23          NaN  2019      4  

[295 rows x 11 columns]

The prices where the amount was zero made the price per unit and price per gram be null too.

In [11]:
# Create new dataframe where there isn't prices with amounts equal to 0
df_noNA = df[~(df['amount'] == 0)]

In [12]:
df_noNA.isnull().sum()

metab_id            0
metab_name        364
price               0
amount              0
unit                0
provider_name       0
source              0
price_per_unit      0
price_per_g         0
year                0
month               0
dtype: int64

New dataframe **df_noNA** doesn't have null prices. It's clean of prices with amounts equal to zero.

In [13]:
df_noNA.count()

metab_id          836593
metab_name        836229
price             836593
amount            836593
unit              836593
provider_name     836593
source            836593
price_per_unit    836593
price_per_g       836593
year              836593
month             836593
dtype: int64

Here we can see that the null values of the price per unit and price per gram columns were erased, however we can also see that the nulls from the metab_name column weren't erased, and that's the reason why there are less rows.

Nevertheless, it's not problematic.

In [14]:
# Dataframe with no NAN and no duplicates!!! 

df_clean = df_noNA.reset_index().drop_duplicates().set_index('date') # drop all duplicates including the index 
# for that we had to convert the index as a column and then put it again as index

df_clean

metab_id                    metab_name  price  amount  \
date                                                                         
2018-01-11 21:39:13         1                    tobramycin  158.0     1.0   
2018-02-11 23:46:53         1                    tobramycin  158.0     1.0   
2018-03-11 23:47:07         1                    tobramycin  158.0     1.0   
2018-04-11 22:47:14         1                    tobramycin  158.0     1.0   
2018-05-11 22:47:27         1                    tobramycin  158.0     1.0   
...                       ...                           ...    ...     ...   
2019-05-06 11:26:13      4128  2-Methyl-5-nitro-2H-indazole  760.0    10.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole   30.0   500.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole   90.0     1.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole  380.0     5.0   
2019-06-14 07:36:43      4128  2-Methyl-5-nitro-2H-indazole  760.0    10.0   

                    unit    provider_name     source  price_per_unit  \
date                                                                   
2018-01-11 21:39:13    g  ACC Corporation  CHEMSPACE          158.00   
2018-02-11 23:46:53    g  ACC Corporation  CHEMSPACE          158.00   
2018-03-11 23:47:07    g  ACC Corporation  CHEMSPACE          158.00   
2018-04-11 22:47:14    g  ACC Corporation  CHEMSPACE          158.00   
2018-05-11 22:47:27    g  ACC Corporation  CHEMSPACE          158.00   
...                  ...              ...        ...             ...   
2019-05-06 11:26:13    g     LeadGen Labs    MolPort           76.00   
2019-06-14 07:36:43   mg     LeadGen Labs    MolPort            0.06   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           90.00   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           76.00   
2019-06-14 07:36:43    g     LeadGen Labs    MolPort           76.00   

                     price_per_g  year  month  
date                                           
2018-01-11 21:39:13    158.00000  2018      1  
2018-02-11 23:46:53    158.00000  2018      2  
2018-03-11 23:47:07    158.00000  2018      3  
2018-04-11 22:47:14    158.00000  2018      4  
2018-05-11 22:47:27    158.00000  2018      5  
...                          ...   ...    ...  
2019-05-06 11:26:13     76.00000  2019      5  
2019-06-14 07:36:43      0.00006  2019      6  
2019-06-14 07:36:43     90.00000  2019      6  
2019-06-14 07:36:43     76.00000  2019      6  
2019-06-14 07:36:43     76.00000  2019      6  

[825408 rows x 11 columns]

In [15]:
df_clean.isnull().sum() 

metab_id            0
metab_name        364
price               0
amount              0
unit                0
provider_name       0
source              0
price_per_unit      0
price_per_g         0
year                0
month               0
dtype: int64

In [16]:
df_clean.count()

metab_id          825408
metab_name        825044
price             825408
amount            825408
unit              825408
provider_name     825408
source            825408
price_per_unit    825408
price_per_g       825408
year              825408
month             825408
dtype: int64

Has it is possible to observe, this new dataframe **df_clean**, besides **not having NaN** values, also **doesn't have duplicates**!

This doesn't change the number of metabolites and providers, but changes the number of prices: 
**825408 preços**

In [17]:
dupl_met                                    #dictionary of duplicated metabolites

{'4-Benzyloxyphenylacetic acid': [1159, 4062],
 'AC1L2SOM': [1204, 3135],
 'Creatine phosphate disodium salt': [546, 3567],
 'Glycyrrhizic acid': [2592, 3708],
 'HYDROXOCOBALAMIN': [2745, 3973],
 'L-cystine': [679, 3367],
 'Pregna-1,4-diene-3,20-dione,16,21-bis(acetyloxy)-9-fluoro-11,17-dihydroxy-, (11b,16a)-': [3004,
  3776],
 'SCHEMBL952350': [2945, 4008],
 'Temsirolimus': [169, 3514],
 'aluminium triphosphate': [3059, 3805],
 'deltamethrin': [624, 2567],
 'pimaricin': [2974, 4196]}

In [18]:
# Change prices of duplicated metabolites so they are set to only one metabolite!

# df_clean1 = df_clean.copy()                                                   # to keep df_clean from being modified
 
for name in dupl_met:                                                         # for each name of the metabolite in dictionary of the duplicates
    id_old = dupl_met[name][1]                                                # save as id_old the id that its gonna change (preferably the more recent id)
    id_new = dupl_met[name][0]                                                # save as id_new the id for which the id_old is gonna change (preferably the older one)
    if name in df_clean.metab_name.unique():                                 # if that metabolite name exists in the dataframe
        df_test = df_clean[df_clean.metab_name == name]                     # create a new df just with the prices for that metabolite name
#         print(df_test)
#         print(name, df_test.metab_id.unique(), id_new)
        if id_new in df_test.metab_id.unique():                               # if the new id exists as an id (this makes sure that if the id does not exist the old one doesn't need to be changed)
#             print(name, df_test.metab_id.unique(), id_new)
            df_clean.loc[df_clean.metab_id == id_old, 'metab_id'] = id_new  # then, change the old one for the new id
#     print(df_clean1[df_clean.metab_name == name])

  
df_clean.metab_id.nunique()

2278

In [19]:
df_clean.provider_name.nunique()

126

The dataframe **df_clean** is now update **without duplicated metabolites**!

It has now 2278 metabolites, 126 providers and 825408 prices.

In [20]:
# as we can see there is only mass units in this dataframe
df_clean['unit'].unique()

array(['g', 'mg', 'kg'], dtype=object)

- **How many providers have prices for all the months?**
**And for 90% of the months?** (regardless of the metabolite)

And what about the providers?? Are they consistently lacking prices for several months independently of the metabolite?

In [21]:
# Calculating how many providers have prices for **all** the months of the time window

dict_dates_prov = {}                     # Create a dictionary that will have {prov_name : [(year,month)]}
t_window = []                            # Create a list of the months that include the time window
for k,g in df_clean.groupby(['provider_name','year','month'])['price_per_g']:  # group by provider_name, year and month 
    k = list(k)                          # Turn to a list so the output is [prov_name, year, month]
#     print(k)
#     print(g)
    num_p = len(g)                       # setting number of prices for each date and each prov_name (for the future maybe)
    date = (k[1],k[2])                   # turn into a tuple as (year, month)   
    if date not in t_window:
        t_window.append(date)            # put on list t_window all months with prices (in other words, all months of the time window)
    if k[0] not in dict_dates_prov:
        dict_dates_prov[k[0]] = [date]
    elif k[0] in dict_dates_prov:
        dict_dates_prov[k[0]].append(date)

        
num_all_providers = len(dict_dates_prov)
# print(dict_dates_prov)

num_prov_all_month = 0                    # count the number of providers that have prices for all the months
prov_all_month = []                       # list of the providers names of those metabolites mentioned above
prov_not_all_m = []                       # list of the providers names that don't have prices for all the months
for name in dict_dates_prov:                    
    if dict_dates_prov[name] == t_window:       # comparing the two lists to see if they are equal!
        prov_all_month.append(name)
        num_prov_all_month += 1
    else: prov_not_all_m.append(name)
        
               
print('Number of providers that have prices for all the months:', num_prov_all_month, '\n')
print('Providers names of the ones that have prices for all the months: \n', prov_all_month, '\n')
print('Percentage:', num_prov_all_month*100/num_all_providers,'%', '(', num_prov_all_month,'of',num_all_providers,'providers) \n')
print('-----------------------------------------------------')
print('Providers names of the ones that don\'t have prices for all months:\n', prov_not_all_m,'\n')

Number of providers that have prices for all the months: 0 

Providers names of the ones that have prices for all the months: 
 [] 

Percentage: 0.0 % ( 0 of 126 providers) 

-----------------------------------------------------
Providers names of the ones that don't have prices for all months:
 ['A1 BioChem Labs', 'AA BLOCKS', 'ACB Blocks', 'ACC Corporation', 'ACINTS', 'ACS Scientific Inc', 'AD Chemical Inc.', 'AK Scientific', 'AK Scientific Inc.', 'AK Scientific, Inc.', 'AKL Research LLP', 'AOBChem USA', 'AbaChemscene', 'Abamachem Ltd', 'Acadechem Co., Ltd', 'Accela ChemBio Inc.', 'Advamacs', 'Aldlab', 'Aldlab Chemicals LLC', 'Alichem', 'Alinda Balt', 'Alinda Chemical Trading Company', 'Alinda Chemical, Ltd.', 'Ambeed, Inc.', 'AnalytiCon Discovery, GmbH', 'Angel Pharmatech Ltd.', 'Angene', 'Angene (China)', 'Angene International Limited', 'Apollo Scientific', 'Arctom Chemicals LLC', 'Ark Pharm, Inc.', 'Aronis', 'Astatech Inc', 'Axon Medchem', 'BCH Research LLP', 'BIONET - Key Organic

In [22]:
# Making sure there is no errors
dict_dates_prov[prov_not_all_m[0]]   

[(2018, 11),
 (2018, 12),
 (2019, 1),
 (2019, 2),
 (2019, 3),
 (2019, 4),
 (2019, 5),
 (2019, 6)]

In [23]:
# And for 90% of the months? 

num_prov_90_month = 0                     # count the number of providers that have prices for at least 90% of the months
prov_90_month = []                        # list of the providers names of those providers mentioned above
prov_not_90_m = []                        # list of the providers names that don't have prices for 90% of the months
len_window = len(t_window)               # there are 18 months so the len will be 18
for name in dict_dates_prov:
    len_name = len(dict_dates_prov[name])        
    if all(date in t_window for date in dict_dates_prov[name]):   # even if the providers doesn't have all the months they need to be contained on the t_window list
        if len_name/len_window >= 0.9:     # find the providers that have prices for 90% of the months by comparing the lens of the lists  
            prov_90_month.append(name)
            num_prov_90_month += 1
        else: prov_not_90_m.append(name)

        
print('Number of providers that have prices for 90% of the months:', num_prov_90_month, '\n')
print('Providers names of the ones that have prices for 90% of the months: \n', prov_90_month, '\n')
print('Percentage:', num_prov_90_month*100/num_all_providers,'%', '(', num_prov_90_month,'of',num_all_providers,'providers) \n')        
print('-----------------------------------------------------')
print('Providers names of the ones that don\'t have prices for 90% of the months:\n', prov_not_90_m,'\n')

Number of providers that have prices for 90% of the months: 72 

Providers names of the ones that have prices for 90% of the months: 
 ['ACINTS', 'ACS Scientific Inc', 'AK Scientific', 'AK Scientific Inc.', 'AK Scientific, Inc.', 'AKL Research LLP', 'AOBChem USA', 'Abamachem Ltd', 'Accela ChemBio Inc.', 'Advamacs', 'Aldlab', 'Alichem', 'Alinda Balt', 'AnalytiCon Discovery, GmbH', 'Angene', 'Angene International Limited', 'Apollo Scientific', 'Aronis', 'Axon Medchem', 'BIOTREND Chemicals, AG', 'BioBlocks', 'Biosynth AG', 'Cayman Europe', 'Chem4Pharma', 'ChemBridge Corp.', 'ChemBridge Corporation', 'ChemDiv, Inc.', 'ChemFaces', 'ChemShuttle', 'Curpys Chemicals', 'Debye Scientific', 'EDASA Scientific', 'ENAMINE Ltd.', 'Enamine Ltd Chemspace partner', 'EvoBlocks, Ltd.', 'Eximed', 'FCH Group', 'FCH Group Chemspace partner', 'Fluorochem', 'Fluorochem Limited', 'HTS Biochemie Innovationen', 'Intermed Ltd', 'Key Organics, Ltd.', 'Labseeker', 'LeadGen Labs', 'Life Chemicals Inc.', 'Manchester O

In [24]:
# Making sure there is no errors
dict_dates_prov[prov_not_90_m[0]]   

[(2018, 11),
 (2018, 12),
 (2019, 1),
 (2019, 2),
 (2019, 3),
 (2019, 4),
 (2019, 5),
 (2019, 6)]

**Question:** What do we do with the providers that don't have prices for 90% of the months consistently? Do we filter them out?

Since we can only make an analysis per metabolite, I choose the **metabolite 67** has study case, since it's a metabolite that is in the category of the metabolites that do not have prices for all months. 

Furthermore, this metabolite does not have duplicates.

In [25]:
df_67 = df_clean[df_clean.metab_id == 67]                  #filter the dataframe where the metabolite id is equal to the study case

In [26]:
df_67

metab_id     metab_name   price  amount unit  \
date                                                                
2018-02-09 09:51:25        67  D-cycloserine   10.00     1.0    g   
2018-02-09 09:51:25        67  D-cycloserine   24.67     1.0    g   
2018-02-09 09:51:25        67  D-cycloserine   18.00     5.0    g   
2018-02-09 09:51:25        67  D-cycloserine   47.60     5.0    g   
2018-02-09 09:51:25        67  D-cycloserine   75.00    25.0    g   
...                       ...            ...     ...     ...  ...   
2019-02-07 14:41:34        67  D-cycloserine  149.82   100.0   mg   
2019-03-07 14:41:44        67  D-cycloserine  149.82   100.0   mg   
2019-04-07 13:41:50        67  D-cycloserine  149.82   100.0   mg   
2019-05-07 13:42:02        67  D-cycloserine  149.82   100.0   mg   
2019-06-14 20:40:50        67  D-cycloserine  149.82   100.0   mg   

                              provider_name     source  price_per_unit  \
date                                                                     
2018-02-09 09:51:25      AK Scientific Inc.  CHEMSPACE         10.0000   
2018-02-09 09:51:25      AK Scientific Inc.  CHEMSPACE         24.6700   
2018-02-09 09:51:25      AK Scientific Inc.  CHEMSPACE          3.6000   
2018-02-09 09:51:25      AK Scientific Inc.  CHEMSPACE          9.5200   
2018-02-09 09:51:25      AK Scientific Inc.  CHEMSPACE          3.0000   
...                                     ...        ...             ...   
2019-02-07 14:41:34  BIOTREND Chemicals, AG    MolPort          1.4982   
2019-03-07 14:41:44  BIOTREND Chemicals, AG    MolPort          1.4982   
2019-04-07 13:41:50  BIOTREND Chemicals, AG    MolPort          1.4982   
2019-05-07 13:42:02  BIOTREND Chemicals, AG    MolPort          1.4982   
2019-06-14 20:40:50  BIOTREND Chemicals, AG    MolPort          1.4982   

                     price_per_g  year  month  
date                                           
2018-02-09 09:51:25    10.000000  2018      2  
2018-02-09 09:51:25    24.670000  2018      2  
2018-02-09 09:51:25     3.600000  2018      2  
2018-02-09 09:51:25     9.520000  2018      2  
2018-02-09 09:51:25     3.000000  2018      2  
...                          ...   ...    ...  
2019-02-07 14:41:34     0.001498  2019      2  
2019-03-07 14:41:44     0.001498  2019      3  
2019-04-07 13:41:50     0.001498  2019      4  
2019-05-07 13:42:02     0.001498  2019      5  
2019-06-14 20:40:50     0.001498  2019      6  

[1856 rows x 11 columns]

In [27]:
# Number of prices for metabolite 67
df_67.shape[0]

1856

In [28]:
# Verify if dataframe as only mass units
df_67.unit.unique()

array(['g', 'mg'], dtype=object)

In [29]:
# How many different providers this metabolite has?
df_67.provider_name.nunique()

46

In [31]:
# How many different sources this metabolite has? 
df_67.source.unique()

array(['CHEMSPACE', 'MolPort', 'OXChem'], dtype=object)

This metabolite, methane (id = 67), has **5662 prices**, counting already only the ones with **mass units** and **without duplicated prices**. Besides, this metabolite has **39 providers** and only **3 sources** (which may be good).

### Boxplot of prices por month 

In [32]:
for k,g in df_67.groupby(['year','month'])['price_per_g']:
    print(list(k))
    print(g)

[2018, 2]
date
2018-02-09 09:51:25    10.000000
2018-02-09 09:51:25    24.670000
2018-02-09 09:51:25     3.600000
2018-02-09 09:51:25     9.520000
2018-02-09 09:51:25     3.000000
                         ...    
2018-02-09 09:51:25     0.000160
2018-02-09 09:51:25    40.000000
2018-02-09 09:51:25    30.400000
2018-02-09 09:51:25    17.600000
2018-02-09 09:51:24     0.001498
Name: price_per_g, Length: 113, dtype: float64
[2018, 3]
date
2018-03-09 11:40:07    10.000000
2018-03-09 11:40:07    24.670000
2018-03-09 11:40:07     3.600000
2018-03-09 11:40:07     9.520000
2018-03-09 11:40:07     3.000000
                         ...    
2018-03-09 11:40:07     0.000160
2018-03-09 11:40:07    40.000000
2018-03-09 11:40:07    30.400000
2018-03-09 11:40:07    17.600000
2018-03-09 11:40:07     0.001498
Name: price_per_g, Length: 123, dtype: float64
[2018, 4]
date
2018-04-09 14:47:29    10.000000
2018-04-09 14:47:29    24.670000
2018-04-09 14:47:29     3.600000
2018-04-09 14:47:29     9.520000
201

In [33]:
# Create boxplot with prices per month
# Create plot with plotly

df_67 = df_67.sort_index()


layout = go.Layout(
    title = 'Boxplot of Methane Prices per Month',
    yaxis = dict(
        title = 'Price (USD)'
    ),
    xaxis = dict(
        title = 'Month'
    )
)


dict_month = {1:'Jan', 2:'Feb', 3:'Mar', 4:'Apr', 5:'May', 6:'Jun', 7:'Jul', 8:'Ago', 9:'Sep', 10:'Oct', 11:'Nov', 12:'Dec'}
    
month_list = []
d_y = {}
for k,g in df_67.groupby(['year','month'])['price_per_g']:
    k = list(k)
    k[1] = dict_month[k[1]]
    k = k[1] +' '+ str(k[0])
    month_list.append(k)
# print(month_list)
    y = g.values
    if k not in d_y:
        d_y[k] = y
    elif k in d_y:
        d_y[k].append(y)
    else: print('There is a key lacking on the y dict.')


def update_plot(signals):#, freq):
    data1 = []
    for s in signals:
        trace1 = go.Box(
            y = d_y[s],
            name =  '{}'.format(s),
            boxpoints='outliers',
            boxmean=True                                       # show mean
        )
    
        data1.append(trace1)
      
    fig = go.Figure(data = data1, layout= layout)
    po.iplot(fig)
    
# data = []
# for s in month_list:
#     trace1 = go.Box(
#         y = d_y[s],
#         name = '{}'.format(s),
#         boxpoints='outliers'
#     )

#     data.append(trace1)


signals = widgets.SelectMultiple(options = month_list, description = 'Date')
#freq = widgets.FloatSlider(min=1, max=20, value=1, description = 'Freq')
widgets.interactive(update_plot, signals = signals)#, freq = freq)




interactive(children=(SelectMultiple(description='Date', options=('Feb 2018', 'Mar 2018', 'Apr 2018', 'May 201…

Como podemos ver a variância muda muito entre os diferentes boxplots de cada mês! 

**??**  já nem sei o que é a variância....

### Boxplot of prices por month and provider

In [34]:
# Create boxplot with prices per month and ** per provider**!!
# Create plot with plotly

df_67 = df_67.sort_index()


layout = go.Layout(
    title = 'Boxplot of Methane Prices per Month and per Provider',
    yaxis = dict(
        title = 'Price (USD)'
    ),
    xaxis = dict(
        title = 'Month'
    )
)

#dicionário com os meses todos para poder substituir o número pela palavra
dict_month = {1:'Jan', 2:'Feb', 3:'Mar', 4:'Apr', 5:'May', 6:'Jun', 7:'Jul', 8:'Ago', 9:'Sep', 10:'Oct', 11:'Nov', 12:'Dec'}

#valores para a opção all_providers
all_p = []    
for k,g in df_67.groupby(['year','month'])['price_per_g']:   # só se agrupa por mês e ano
    k = list(k)                                                # o mês e o ano têm de ficar como lista
    k[1] = dict_month[k[1]]                                    # substituir o nº do mês pela palavra
    date = k[1] +' '+ str(k[0])                                # criar a string date que vai ficar no x axis
    y = g.values                                           
    for i in y:                                                # colocar numa lista cada valor com a respectiva data      
        all_p.append([date, i])

dict_boxplot = {} #Fazer um dicionário onde os values são listas de listas (ex., [date, valor])
          
dict_boxplot['all_providers'] = all_p                          # colocar os valores de all_providers no dicionário
for k,g in df_67.groupby(['year','month', 'provider_name'])['price_per_g']:  # agrupar por mês, ano e provider
    k = list(k)                                                # mês, ano e provider têm de ficar como lista para ser mais fácil usar
    k[1] = dict_month[k[1]]                                                        
    date = k[1] +' '+ str(k[0]) 
    y = g.values
    values =[]                                                 # criar lista onde vão ficar as listas 
    for i in y:
        values.append([date, i])                               # pôr na lista uma lista com um dos valores e a respectiva data
#     print(values)
    if k[2] not in dict_boxplot:
        dict_boxplot[k[2]] = values                            # colocar no dicionário o provider e a respectiva lista de listas
    elif k[2] in dict_boxplot:
        dict_boxplot[k[2]].extend(values)                       # se já existir extender a lista existente com outras listas

# print(dict_boxplot)        


list_options=[]                                                # criar uma lista que tenha todos os provviders + all_providers    
for p in dict_boxplot:                                         # para poder usar como opção
    list_options.append(p)

# list_options.sort(key=str.lower)    
        
# print(len(list_options))  #dá certinho com o número de providers que este metabolito tem + a opção de todos os providers    
    
def update_plot(signals):#, freq):
    data1 = []
    for s in signals:
        trace1 = go.Box(
            y = [values[1] for values in dict_boxplot[s]],     # criar gigante lista só com os valores de y (usar valores nas listas da lista)
            x = [values[0] for values in dict_boxplot[s]],     # criar gigante lista com as datas respectivas para cada y
            name =  '{}'.format(s),
            boxpoints= 'all',                                  # 'outliers' assim mostra só os outliers
            boxmean=True                                       # show mean
        )
    
        data1.append(trace1)
      
    fig = go.Figure(data = data1, layout= layout)
    po.iplot(fig)
    

signals = widgets.SelectMultiple(options = list_options, description = 'Date')
#freq = widgets.FloatSlider(min=1, max=20, value=1, description = 'Freq')
widgets.interactive(update_plot, signals = signals)#, freq = freq)

interactive(children=(SelectMultiple(description='Date', options=('all_providers', 'AD Chemical Inc.', 'AK Sci…

In [35]:

#dictionary with all the months to substitute the number by the word
dict_month = {1:'Jan', 2:'Feb', 3:'Mar', 4:'Apr', 5:'May', 6:'Jun', 7:'Jul', 8:'Ago', 9:'Sep', 10:'Oct', 11:'Nov', 12:'Dec'}

time_window = ['Jan 2018', 'Feb 2018', 'Mar 2018', 'Apr 2018', 'May 2018', 'Jun 2018', 'Jul 2018', 'Ago 2018', 'Sep 2018',
               'Oct 2018', 'Nov 2018', 'Dec 2018', 'Jan 2019', 'Feb 2019', 'Mar 2019', 'Apr 2019', 'May 2019', 'Jun 2019']


# list of all the months on this metabolite
time_window_m = [] 
for k,g in df_67.groupby(['year','month'])['price_per_g']:    # só se agrupa por mês e ano
    k = list(k)                                                # o mês e o ano têm de ficar como lista
    k[1] = dict_month[k[1]]                                    # substituir o nº do mês pela palavra
    date = k[1] +' '+ str(k[0])                                # criar a string date que vai ficar no x axis
    if date not in time_window_m:
        time_window_m.append(date)


def date_str(k):    # k is the list that is created when we do the groupby (it has 3 elements: year, month, provider_name)
    k[1] = dict_month[k[1]]                                                        
    date = k[1] +' '+ str(k[0]) 
    return date


dict_prov = {}                                          # Create dictionary with list of lists {prov:[[date,price]]}
for k,g in df_67.groupby(['year','month', 'provider_name'])['price_per_g']:  # group by month, year and provider
    k = list(k)                                         # month, year and provider turn into a list to be easier to use
    date = date_str(k)                                  # turn date into a string 
    y = g.values
    values =[]                                          # create list where it will be the lists  
    for i in y:
        values.append([date, i])                        # put on the list 'values' a list with the date and each price [date, price]
#     print(values)
    if k[2] not in dict_prov:
        dict_prov[k[2]] = values                        # put on the dict the provider as key and the respective list of lists as values
    elif k[2] in dict_prov:
        dict_prov[k[2]].extend(values)                  # if provider already exists, extend the existing list with other lists

# print(dict_prov)
len_all_prov = len(dict_prov)                           # number of all the providers for this metabolite
len_window_m = len(time_window_m)                       # number of all the months with prices for this metabolite
        
prov_all_month = []                                     # create list with providers that have prices for all months
prov_almost_all_m = []                                  # create list with providers that have prices for all months but all the prices are below 0.5 USD
prov_flaw = []                                          # create dictionary as {provider: num_month_without_price}
prov_flaw_below_1 = []                                  # create list with providers without prices for all months and are below 0.5 USD
for prov in dict_prov:
    month = {}                                          # create dictionary as {date:[prices]}
    for values in dict_prov[prov]:                      # each list [date, price]
        if values[0] not in month:                      # if date isn't in the dict put it as key and price in list as value      
            month[values[0]] = [values[1]]
        elif values[0] in month:                        # if date is already in dict, append the price to the list of that date 
            month[values[0]].append(values[1])
#     print(month)
    if len(month) < len_window_m:                       # if len(month) is less than the number of months in this metabolite
        if any(v >= 0.5 for m in month for v in month[m]):
            prov_flaw.append(prov)                      # and if some prices are above 0.5 USD then save it on list of provider lacking months
        elif all(v > 0 and v < 0.5 for m in month for v in month[m]): # if prices are all between 0 and 0.5 USD 
            prov_flaw_below_1.append(prov)              # the provider is kept on the list below 1USD since the prices are really low
    elif len(month) == len_window_m:                    # if they have all the months in this metabolite then verify if some prices are above 0.5 USD
        if any(v >= 0.5 for m in month for v in month[m]):
            prov_all_month.append(prov)                 # if they are then save prov on the list of all months
        elif all(v > 0 and v < 0.5 for m in month for v in month[m]): # if it has 18 months but prices are all between 0 and 0.5 USD 
            prov_almost_all_m.append(prov)              # provider it's kept on list of the almost all months, since prices are too low
                    


# Find which provider doesn't have prices for all the months and how many months are lacking
# There are 18 months on total

dic_prov_flaw = {}                                           # create a dictionary as {provider:num_months_lacking}
for prov in dict_prov:                                       # for each provider
    month = {}                                               # create a new dictionary 'month' as {date:[prices]}
    for values in dict_prov[prov]:                           # for each list of [date, price]
        if values[0] not in month:                           
            month[values[0]] = [values[1]]                   # put the date as key in 'month' and the price in a list as value
        elif values[0] in month:
            month[values[0]].append(values[1])               # if the date already exists on the 'month' just append more prices 
#     print(month)
    for t in time_window_m:                                  # for each date on the list 'time_window_m'
        if t not in month.keys():                            # if the date is not on 'month'
            times = 1                                        # create a counter 'times' equal to 1
            if prov not in dic_prov_flaw:                   
                dic_prov_flaw[prov] = times                  # set the provider as a key in the dict 'dic_prov_flaw' with the value as the counter 'times'
            elif prov in dic_prov_flaw:
                dic_prov_flaw[prov] += 1                     # if the provider already exist on 'dic_prov_flaw' add 1 to the counter on the value
              
        
        
        
print('Providers with prices for each month:\n\n', prov_all_month, '\n')     #providers que têm preços todos os meses!!
print('Percentage: ', len(prov_all_month)*100/len_all_prov,'% (', len(prov_all_month),'of',len_all_prov, ' providers) \n')
print('----------------------------------------------------------------------------------\n')
print('Providers with prices for each month but below 0.5:\n\n', prov_almost_all_m, '\n')
print('Percentage: ', len(prov_almost_all_m)*100/len_all_prov,'% (', len(prov_almost_all_m),'of', len_all_prov,' providers) \n')
print('----------------------------------------------------------------------------------\n')
print('Providers which do not have prices for all months:\n\n', prov_flaw, '\n')
print('Percentage: ', len(prov_flaw)*100/len_all_prov,'% (', len(prov_flaw),'of', len_all_prov, ' providers) \n\n')
print('----------------------------------------------------------------------------------\n')
print('Providers who don\'t have prices for all the months and prices are also consistently below 0.5:\n\n', prov_flaw_below_1,'\n')
print('Percentage: ', len(prov_flaw_below_1)*100/len_all_prov,'% (', len(prov_flaw_below_1),'of', len_all_prov,' providers) \n')
print('----------------------------------------------------------------------------------\n')
print('Which providers do not have prices for all the months and how many months are lacking: \n\n', dic_prov_flaw)

Providers with prices for each month:

 ['AK Scientific Inc.', 'AK Scientific, Inc.', 'Angene', 'Angene International Limited', 'Apollo Scientific', 'Biosynth AG', 'ENAMINE Ltd.', 'Enamine Ltd Chemspace partner', 'Pharmeks, Ltd.', 'TargetMol'] 

Percentage:  21.73913043478261 % ( 10 of 46  providers) 

----------------------------------------------------------------------------------

Providers with prices for each month but below 0.5:

 ['BIOTREND Chemicals, AG', 'Selleck Chemicals LLC'] 

Percentage:  4.3478260869565215 % ( 2 of 46  providers) 

----------------------------------------------------------------------------------

Providers which do not have prices for all months:

 ['AD Chemical Inc.', 'Acadechem Co., Ltd', 'Ark Pharm, Inc.', 'Astatech Inc', 'CARBONE SCIENTIFIC CO., LTD', 'Cohnchem Scientific Co., Ltd', 'Curpys Chemicals', 'Fluorochem', 'Glentham Life Sciences', 'INTERBIOSCREEN Limited', 'InterBioScreen Ltd.', 'Labseeker', 'OXChem', 'Oxchem Corporation', 'StruChem CO.,

There are really few providers with prices for all months in these metabolite. For the providers that don't have prices for all the months we will **keep only the ones that have prices for at least 90% of the months**.

In [36]:
# Which providers of the metabolite 1755 have prices for 90% of the months

dict_prov

{'AD Chemical Inc.': [['Feb 2018', 9.0],
  ['Feb 2018', 3.6],
  ['Feb 2018', 3.5],
  ['Feb 2018', 2.88],
  ['Feb 2018', 2.11],
  ['Mar 2018', 2.11],
  ['Mar 2018', 3.6],
  ['Mar 2018', 9.0],
  ['Mar 2018', 2.88],
  ['Mar 2018', 3.5]],
 'AK Scientific Inc.': [['Feb 2018', 10.0],
  ['Feb 2018', 2.42],
  ['Feb 2018', 3.8939999999999997],
  ['Feb 2018', 24.67],
  ['Feb 2018', 5.052],
  ['Feb 2018', 3.0],
  ['Feb 2018', 9.52],
  ['Feb 2018', 3.6],
  ['Mar 2018', 10.0],
  ['Mar 2018', 9.52],
  ['Mar 2018', 3.0],
  ['Mar 2018', 5.052],
  ['Mar 2018', 24.67],
  ['Mar 2018', 2.42],
  ['Mar 2018', 3.8939999999999997],
  ['Mar 2018', 3.6],
  ['Apr 2018', 9.52],
  ['Apr 2018', 3.0],
  ['Apr 2018', 5.052],
  ['Apr 2018', 2.42],
  ['Apr 2018', 3.8939999999999997],
  ['Apr 2018', 3.6],
  ['Apr 2018', 24.67],
  ['Apr 2018', 10.0],
  ['May 2018', 5.052],
  ['May 2018', 2.42],
  ['May 2018', 3.8939999999999997],
  ['May 2018', 24.67],
  ['May 2018', 10.0],
  ['May 2018', 3.6],
  ['May 2018', 9.52],
  ['

In [37]:
# And for 90% of the months? # colocar media num dicionario com datas e medias

prov_90 = []                             # list of providers that have prices for 90% of the months
prov_allm = []                           # list of providers that have prices for all months
len_window_m = len(time_window_m)        # number of months on this metabolite
prov_not_90 = []                         # list of the providers names that don't have prices for 90% of the months
for prov in dict_prov:                   # this dictionary has all providers {provider:[[date, price]]}
    dates = {}                           # create dictionary where {date:[prices]}
    for d in dict_prov[prov]:
        if d[0] not in dates:
            dates[d[0]] = [d[1]]
        elif d[0] in dates:
            dates[d[0]].append(d[1])
    len_dates = len(dates)                      # find number of months with prices for that provider
    if all(date in time_window_m for date in dates):   # even if the providers don't have all the months they need to be contained on the t_window list
        if len_dates/len_window_m == 1:         # find providers that have prices for all months
            prov_allm.append(prov)
        elif len_dates/len_window_m >= 0.9:     # find the providers that have prices for 90% of the months by comparing the lens of the lists  
            prov_90.append(prov)
        else: prov_not_90.append(prov)

            
print('Providers that have prices for all months: \n', prov_allm, '\n')
print('Percentage:', len(prov_allm)*100/len(dict_prov),'%', '(', len(prov_allm),'of',len(dict_prov),'providers) \n')        
print('-----------------------------------------------------')     
print('Providers that have prices for 90% of the months: \n', prov_90, '\n')
print('Percentage:', len(prov_90)*100/len(dict_prov),'%', '(', len(prov_90),'of',len(dict_prov),'providers) \n')        
print('-----------------------------------------------------')
print('Providers that don\'t have prices for 90% of the months:\n', prov_not_90,'\n')
print('Percentage:', len(prov_not_90)*100/len(dict_prov),'%', '(', len(prov_not_90),'of',len(dict_prov),'providers) \n')

Providers that have prices for all months: 
 ['AK Scientific Inc.', 'AK Scientific, Inc.', 'Angene', 'Angene International Limited', 'Apollo Scientific', 'BIOTREND Chemicals, AG', 'Biosynth AG', 'ENAMINE Ltd.', 'Enamine Ltd Chemspace partner', 'Pharmeks, Ltd.', 'Selleck Chemicals LLC', 'TargetMol'] 

Percentage: 26.08695652173913 % ( 12 of 46 providers) 

-----------------------------------------------------
Providers that have prices for 90% of the months: 
 ['labseeker Inc'] 

Percentage: 2.1739130434782608 % ( 1 of 46 providers) 

-----------------------------------------------------
Providers that don't have prices for 90% of the months:
 ['AD Chemical Inc.', 'Acadechem Co., Ltd', 'Ark Pharm, Inc.', 'Astatech Inc', 'BIONET/Key Organics Ltd.', 'Bide Pharmatech Ltd.', 'CARBONE SCIENTIFIC CO., LTD', 'Cohnchem Scientific Co., Ltd', 'Curpys Chemicals', 'Fluorochem', 'Glentham Life Sciences', 'INTERBIOSCREEN Limited', 'InterBioScreen Ltd.', 'Labseeker', 'OXChem', 'Oxchem Corporation', 'S

**ATENTION** Here the providers that have prices for 90% of the months don't include the providers with prices for all months (in contrary to the percentages on the initial statistics)

In [38]:
def turn_month_in_list(month):
    month = month.split()
    for number, word in dict_month.items():
        if word == month[0]:
            month[0] = number
    month[1] = int(month[1])
    return month

In [47]:
def find_previous_month(month):
    for i in range(len(time_window_m)):
        if time_window_m[i]== month:
            if i-1 < 0:
                print('There is no previous month for this one. \n')
            else: 
                prev_month = time_window_m[i-1]    
                return prev_month
    

In [40]:
def avg_prev_month(previous_month, dic_fix):
    if previous_month in dic_fix:
        avg = 0
        for date in dic_fix:                                    # for each date of this provider
            if date == previous_month:                          # see if it is equal to the previous month of the lacking one                                               
                sum = 0              
                for p in dic_fix[date]:
                    sum += p                                    # sum the prices of that previous month  
                avg = sum/len(dic_fix[date])                    # and calculate the average
        return avg        
    else: 
        previous_month = find_previous_month(previous_month)    # find previous month of the previous month
        avg = avg_prev_month(previous_month)
        

In [46]:
# Create a function to rectify one provider at a time

def rectify_prov(prov, df):
    
    # Discover which month is lacking to the provider that needs to be rectified
    
    time_fix = []                                             # create list with the dates that will have to be amend
    dic_fix = {}                                              # dictionary with {date:[prices]} for that only provider
    time = []                                                 # create list with all the dates of this provider
    for value in dict_prov[prov]:                             # lists of [date, price]
        if value[0] not in dic_fix:                           # create dic_fix
            dic_fix[value[0]] = [value[1]]
        elif value[0] in dic_fix:
            dic_fix[value[0]].append(value[1])
    for date_m in dic_fix:                                    # for each time in the dictionary with the dates 
        if date_m not in time:
            time.append(date_m)
    for t in time_window_m:                                   # for each date in the time window 
        if t not in time:
            time_fix.append(t)

#     print('Months lacking prices:', time_fix)


    # Rectify the provider!

    avg_month = {}                                                  # create dictionary as {month:avg}
    for time in time_fix:                                           # for each date that needes to be fixed
        previous_month = find_previous_month(time)                  # find previous month of the lacking one
        if previous_month:
            avg = avg_prev_month(previous_month, dic_fix)
            if time not in avg_month:                                   
                avg_month[time] = avg                               # set time as a key on 'avg_month' and the average as value 
            elif time in avg_month:
                print('Atention: There is more than 1 average for a month. \n')
        else: 
            print('It\'s not possible to rectify the provider', prov, '\n')
            break
            
    print('Average of the previous month:', avg_month, '\n')
    
    #**These code needs to be optimizes so it can be used if the provider is lacking more than one month or various months in a row!**

    for month in avg_month:
        previous_month = find_previous_month(month)            # find previous month of 'month' 
        month = turn_month_in_list(month)                      # turn 'Month Year' into [month_number, year]    
        p_month = turn_month_in_list(previous_month)           
        last_index = df[(df['year'] == month[1]) &
                        (df['month'] == month[0])].index.max() # save the last index of the df for that month and year
        
        last_index = last_index + datetime.timedelta(minutes=30)         # add thirty minutes to the last_index to create a new index

        df_fix = df[(df.provider_name == prov) &
                    (df['year'] == p_month[1]) &
                    (df['month'] == p_month[0])].tail(1)         # save the last row of the df for that provider in the previous month
        
#         print(df_fix)
        
        metab_id, metab_name, provider_name, source = df_fix.loc[df_fix.index.max(),
                                                                 ['metab_id','metab_name',
                                                                  'provider_name','source']
                                                                ]
        # set the metab_id, metab_name, provider_name and source as the last values for that provider in the previous 
        # month so we can use them below
        
        amount = 1                                               # set amount and unit always as 1g so it doesn't mess up on
        unit = 'g'                                                 # the price_per_unit and price_per_g
        
        df_fix= pd.DataFrame([[metab_id,metab_name, avg, amount, unit, provider_name,source, avg, avg, month[1], month[0]]],
                             columns=list(df.columns), index = [last_index])
        # create a new dataframe with just one row with all the right values and the average 
        
#         print(df_fix)

        df = df.append(df_fix.loc[:])                          # add this row to the dataframe
        return df

In [48]:
# df_67_copy = df_67.copy()

for prov in prov_90:
    df_67= rectify_prov(prov, df_67)

df_67

There is no previous month for this one. 

It's not possible to rectify the provider labseeker Inc 

Average of the previous month: {} 



**ATENCAO: Este código em cima precisa URGENTEMENTE de ser optimizado para poder ser utilizado quase haja mais providers com falhas em meses, haja mais meses em falta e/ou os meses em falta estejam seguidos**

In [ ]:
# df_67.index.max()                  # watch if the max index changed as it should

In [ ]:
# Clean the providers that don't have prices for at least 90% of the months 

df_67 = df_67[~df_67.provider_name.isin(prov_not_90)] # save in the database only the data related to the providers that 
                                                            # have prices for at least 90% of the months
    
df_67